syntax = "proto3";

package mdining;

import "google/api/annotations.proto";
import "proto/dininghalls.proto";
import "proto/items.proto";
import "proto/filterableentries.proto";
import "proto/menu.proto";
import "proto/food.proto";
import "proto/foodstat.proto";
import "proto/hearts.proto";

service MDining {
  rpc GetDiningHalls(DiningHallsRequest) returns (DiningHallsReply) {
    option (google.api.http) = {
      get : "/v1/diningHalls"
    };
  }

  rpc GetItems(ItemsRequest) returns (ItemsReply) {
    option (google.api.http) = {
      get : "/v1/items"
    };
  }

  rpc GetFilterableEntries(FilterableEntriesRequest)
      returns (FilterableEntriesReply) {
    option (google.api.http) = {
      get : "/v1/filterableEntries"
    };
  }

  rpc GetAll(AllRequest) returns (AllReply) {
    option (google.api.http) = {
      get: "/v1/all"
    };
  }

  rpc GetMenu(MenuRequest) returns (MenuReply) {
    option (google.api.http) = {
      get: "/v1/menus"
    };
  }

  rpc GetFood(FoodRequest) returns (FoodReply) {
    option (google.api.http) = {
      get: "/v1/foods"
    };
  }

  rpc GetFoodStats(FoodStatsRequest) returns (FoodStatsReply) {
    option (google.api.http) = {
      get: "/v1/stats"
    };
  }

  rpc GetSummaryStats(SummaryStatsRequest) returns (SummaryStatsReply) {
    option (google.api.http) = {
      get: "/v1/summarystats"
    };
  }

  rpc StreamHearts(HeartsRequest) returns (stream HeartsReply) {
  }

  rpc GetHearts(HeartsRequest) returns (HeartsReply) {
    option (google.api.http) = {
      get: "/v1/hearts"
    };
  }

  rpc AddHeart(HeartsRequest) returns (HeartsReply) {
    option (google.api.http) = {
      post: "/v1/hearts"
    };
  }
}

message FoodStatsRequest {}
message FoodStatsReply { repeated FoodStat foodStats = 1; }

message SummaryStatsRequest {
  string startDate = 1;
  string endDate = 2;
}
message SummaryStatsReply {
  SummaryStats stats = 1;
}

message DiningHallsRequest {}

message DiningHallsReply { repeated DiningHall diningHalls = 1; }

message ItemsRequest {}

message ItemsReply { map<string, Item> items = 1; }

message FilterableEntriesRequest {}

message FilterableEntriesReply { repeated FilterableEntry filterableEntries = 1; }

message AllRequest {}

message AllReply {
  repeated DiningHall diningHalls = 1;
  map<string, Item> items = 2;
  repeated FilterableEntry filterableEntries = 3;
}

message MenuRequest {
  string date = 1;
  string meal = 2;
  string diningHall = 3;
}

message MenuReply { repeated Menu menus = 1; }

message FoodRequest {
  string name = 1;
  string date = 2;
  string startDate = 3;
  string endDate = 4;
}

message FoodReply {
  repeated Food foods = 1;
}

message HeartsRequest {
  repeated string keys = 1;
}

message HeartsReply {
  repeated HeartCount counts = 1;
}
